File: app/src/main/java/me/jfenn/alarmio/fragments/sound/RadioSoundChooserFragment.java
package me.jfenn.alarmio.fragments.sound;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.URLUtil;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatButton;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import me.jfenn.alarmio.R;
import me.jfenn.alarmio.adapters.SoundsAdapter;
import me.jfenn.alarmio.data.SoundData;
import me.jfenn.alarmio.fragments.BasePagerFragment;
import me.jfenn.alarmio.interfaces.SoundChooserListener;

public class RadioSoundChooserFragment extends BaseSoundChooserFragment {

    private static final String SEPARATOR = ":AlarmioRadioSound:";
    private static final String PREF_RADIOS = "previousRadios";

    private EditText radioUrlEditText;
    private AppCompatButton testRadio;
    private SoundData currentSound;

    private SharedPreferences prefs;
    private List<SoundData> sounds;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_sound_chooser_radio, container, false);

        prefs = PreferenceManager.getDefaultSharedPreferences(getContext());

        radioUrlEditText = view.findViewById(R.id.radioUrl);
        TextView errorTextView = view.findViewById(R.id.errorText);
        RecyclerView recycler = view.findViewById(R.id.recycler);
        testRadio = view.findViewById(R.id.testRadio);

        List<String> previousRadios = new ArrayList<>(prefs.getStringSet(PREF_RADIOS, new HashSet<String>()));
        Collections.sort(previousRadios, (o1, o2) -> {
            try {
                return Integer.parseInt(o1.split(SEPARATOR)[0]) - Integer.parseInt(o2.split(SEPARATOR)[0]);
            } catch (NumberFormatException e) {
                return 0;
            }
        });

        sounds = new ArrayList<>();
        for (String string : previousRadios) {
            String url = string.split(SEPARATOR)[1];
            sounds.add(new SoundData(url, SoundData.TYPE_RADIO, url));
        }

        recycler.setLayoutManager(new LinearLayoutManager(getContext()));

        SoundsAdapter adapter = new SoundsAdapter(getAlarmio(), sounds);
        adapter.setListener(this);
        recycler.setAdapter(adapter);

        testRadio.setOnClickListener(v -> {
            if (currentSound == null) {
                currentSound = new SoundData("", SoundData.TYPE_RADIO, radioUrlEditText.getText().toString());
                try {
                    currentSound.preview(getAlarmio());
                    testRadio.setText(R.string.title_radio_stop);
                } catch (Exception e) {
                    e.printStackTrace();
                    Toast.makeText(getContext(), e.getMessage(), Toast.LENGTH_SHORT).show();
                }
            } else {
                if (currentSound.isPlaying(getAlarmio()))
                    currentSound.stop(getAlarmio());

                currentSound = null;
                testRadio.setText(R.string.title_radio_test);
            }
        });

        view.findViewById(R.id.createRadio).setOnClickListener(v -> {
            if (URLUtil.isValidUrl(radioUrlEditText.getText().toString())) {
                onSoundChosen(new SoundData(getString(R.string.title_radio), SoundData.TYPE_RADIO, radioUrlEditText.getText().toString()));
            }
        });

        return view;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        radioUrlEditText = null;
        testRadio = null;
    }

    @Override
    public void onSoundChosen(SoundData sound) {
        super.onSoundChosen(sound);

        if (sound != null) {
            if (sounds.contains(sound))
                sounds.remove(sound);

            sounds.add(0, sound);

            Set<String> radios = new HashSet<>();
            for (int i = 0; i < sounds.size(); i++) {
                radios.add(i + SEPARATOR + sounds.get(i).getUrl());
            }

            prefs.edit().putStringSet(PREF_RADIOS, radios).apply();
        }
    }

    @Override
    public String getTitle(Context context) {
        return context.getString(R.string.title_radio);
    }

    public static class Instantiator extends BaseSoundChooserFragment.Instantiator {

        public Instantiator(Context context, SoundChooserListener listener) {
            super(context, listener);
        }

        @Override
        BasePagerFragment newInstance(int position, SoundChooserListener listener) {
            BaseSoundChooserFragment fragment = new RadioSoundChooserFragment();
            fragment.setListener(listener);
            return fragment;
        }

        @Override
        public String getTitle(Context context, int position) {
            return context.getString(R.string.title_radio);
        }
    }

}

File: app/src/main/java/me/jfenn/alarmio/fragments/sound/RadioSoundChooserFragment.kt
package me.jfenn.alarmio.fragments.sound

import android.content.Context
import android.content.SharedPreferences
import android.os.Bundle
import android.preference.PreferenceManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.webkit.URLUtil
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.widget.AppCompatButton
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import me.jfenn.alarmio.R
import me.jfenn.alarmio.adapters.SoundsAdapter
import me.jfenn.alarmio.data.SoundData
import me.jfenn.alarmio.fragments.BasePagerFragment
import me.jfenn.alarmio.interfaces.SoundChooserListener
import me.jfenn.alarmio.interfaces.SoundPlayer
import org.koin.android.ext.android.inject
import java.util.*

class RadioSoundChooserFragment : BaseSoundChooserFragment() {

    companion object {
        private const val SEPARATOR = ":AlarmioRadioSound:"
        private const val PREF_RADIOS = "previousRadios"
    }

    private var radioUrlEditText: EditText? = null
    private var testRadio: AppCompatButton? = null

    private val prefs: SharedPreferences by lazy {
        PreferenceManager.getDefaultSharedPreferences(context)
    }

    private var sounds: MutableList<SoundData>? = null

    private val player: SoundPlayer by inject()
    private var isPlaying: Boolean = false

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_sound_chooser_radio, container, false)
        radioUrlEditText = view.findViewById(R.id.radioUrl)
        val errorTextView = view.findViewById<TextView>(R.id.errorText)
        val recycler: RecyclerView = view.findViewById(R.id.recycler)
        testRadio = view.findViewById(R.id.testRadio)

        val previousRadios = ArrayList(prefs.getStringSet(PREF_RADIOS, HashSet())!!)

        previousRadios.sortWith(Comparator { o1: String, o2: String ->
            try {
                o1.split(SEPARATOR)[0].toInt() - o2.split(SEPARATOR)[0].toInt()
            } catch (e: NumberFormatException) {
                0
            }
        })

        sounds = ArrayList()
        for (string in previousRadios) {
            val url = string.split(SEPARATOR).toTypedArray()[1]
            sounds?.add(SoundData(url, SoundData.TYPE_RADIO, url))
        }

        recycler.layoutManager = LinearLayoutManager(context)
        val adapter = SoundsAdapter(alarmio, sounds, player)
        adapter.setListener(this)
        recycler.adapter = adapter

        testRadio?.setOnClickListener {
            val currentSound = SoundData("", SoundData.TYPE_RADIO, radioUrlEditText?.text?.toString() ?: "")
            player.stop()

            if (isPlaying) {
                testRadio?.setText(R.string.title_radio_test)
                isPlaying = false
            } else {
                try {
                    player.play(currentSound)
                } catch (e: Exception) {
                    e.printStackTrace()
                    Toast.makeText(context, e.message, Toast.LENGTH_SHORT).show()
                }

                testRadio?.setText(R.string.title_radio_stop)
                isPlaying = true
            }
        }

        view.findViewById<View>(R.id.createRadio).setOnClickListener {
            radioUrlEditText?.text?.toString()?.let { url ->
                if (URLUtil.isValidUrl(url))
                    onSoundChosen(SoundData(getString(R.string.title_radio), SoundData.TYPE_RADIO, url))
            }
        }

        return view
    }

    override fun onDestroyView() {
        super.onDestroyView()
        radioUrlEditText = null
        testRadio = null
    }

    override fun onSoundChosen(sound: SoundData?) {
        super.onSoundChosen(sound)

        sounds?.let { sounds ->
            sound?.let { // add sound (it) to top of list
                if (sounds.contains(it))
                    sounds.remove(it)

                sounds.add(0, it)
            }

            // store sounds in SharedPreferences
            sounds.mapIndexed { index, sound ->
                "${index}${SEPARATOR}${sound.url}"
            }.let { files ->
                prefs.edit().putStringSet(PREF_RADIOS, files.toHashSet()).apply()
            }
        }
    }

    override fun getTitle(context: Context): String {
        return context.getString(R.string.title_radio)
    }

    class Instantiator(context: Context?, listener: SoundChooserListener?) : BaseSoundChooserFragment.Instantiator(context, listener) {
        public override fun newInstance(position: Int, listener: SoundChooserListener): BasePagerFragment {
            val fragment: BaseSoundChooserFragment = RadioSoundChooserFragment()
            fragment.setListener(listener)
            return fragment
        }

        override fun getTitle(context: Context, position: Int): String {
            return context.getString(R.string.title_radio)
        }
    }
}